src/arch/riscv/irqc/aia/aplic.c

```diff
@@ -0,0 +1,237 @@
+/**
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) Bao Project and Contributors. All rights reserved.
+ */
+
+#include <aplic.h>
+#include <cpu.h>
+#include <interrupts.h>
+
+/** APLIC fields and masks defines */
+#define APLIC_DOMAINCFG_CTRL_MASK       (0x1FF)
+
+#define DOMAINCFG_DM                    (1U << 2)
+
+#define INTP_IDENTITY                   (16)
+#define INTP_IDENTITY_MASK              (0x3FF)
+
+#define APLIC_DISABLE_IDELIVERY	        (0)
+#define APLIC_ENABLE_IDELIVERY	        (1)
+#define APLIC_DISABLE_IFORCE	        (0)
+#define APLIC_ENABLE_IFORCE	            (1)
+#define APLIC_IDC_ITHRESHOLD_EN_ALL     (0)
+#define APLIC_IDC_ITHRESHOLD_DISBL_ALL  (1)
+
+#define APLIC_SRCCFG_DEFAULT            APLIC_SRCCFG_DETACH    
+
+/** APLIC public data */
+volatile struct irqc_global_hw *irqc_global;
+volatile struct irqc_hart_hw *irqc_hart;
+
+/** APLIC private data */
+size_t APLIC_IMPL_INTERRUPTS;
+size_t APLIC_IMPL_INTERRUPTS_REGS;
+uint32_t impl_src[APLIC_MAX_INTERRUPTS];
+
+/** APLIC private functions */
+/**
+ * @brief Populate impl_src array with 1's if source i is an active
+ *        source in this domain
+ * 
+ * @return size_t total number of implemented interrupts
+ */
+static size_t aplic_scan_impl_int(void)
+{
+    size_t count = APLIC_MAX_INTERRUPTS-1;
+    for (size_t i = 0; i < APLIC_MAX_INTERRUPTS-1; i++) {
+        irqc_global->sourcecfg[i] = APLIC_SOURCECFG_SM_DEFAULT;
+        impl_src[i] = IMPLEMENTED;
+        if (irqc_global->sourcecfg[i] == APLIC_SOURCECFG_SM_INACTIVE) {
+            impl_src[i] = NOT_IMPLEMENTED;
+            count -= 1;           
+        }
+        irqc_global->sourcecfg[i] = APLIC_SOURCECFG_SM_INACTIVE;
+    }
+    return count;
+}
+
+/** APLIC public functions */
+/** Initialization Functions */
+void aplic_init(void)
+{
+    irqc_global->domaincfg = 0;
+
+    /** Clear all pending and enabled bits*/
+    for (size_t i = 0; i < APLIC_NUM_CLRIx_REGS; i++) {
+        irqc_global->setip[i] = 0;
+        irqc_global->setie[i] = 0;
+    }
+
+    APLIC_IMPL_INTERRUPTS = aplic_scan_impl_int();
+
+    /** Sets the default value of hart index and prio for implemented sources*/
+    for (size_t i = 0; i < APLIC_NUM_TARGET_REGS; i++){
+        if(impl_src[i] == IMPLEMENTED){
+            irqc_global->target[i] = APLIC_TARGET_PRIO_DEFAULT;
+        }
+    }
+
+    irqc_global->domaincfg |= APLIC_DOMAINCFG_IE;
+}
+
+void aplic_idc_init(void){
+    uint32_t idc_index = cpu()->id;
+    irqc_hart[idc_index].ithreshold = APLIC_IDC_ITHRESHOLD_EN_ALL;  
+    irqc_hart[idc_index].idelivery = APLIC_ENABLE_IDELIVERY;
+    irqc_hart[idc_index].iforce = APLIC_DISABLE_IFORCE;
+}
+
+/** Domain functions */
+void aplic_set_sourcecfg(irqid_t int_id, uint32_t val)
+{
+    uint32_t real_int_id = int_id - 1;
+    if(impl_src[real_int_id] == IMPLEMENTED){
+        if(!(val & APLIC_SRCCFG_D) && 
```

Comment:
I believe you are saying D = 0 (interrupt not delegated)

Reply:
If that is the case, I complete agree with you. Also because the way the vaplic is made val will never have D = 1. It was a bad decision on my part.

Reply:
Now that I think better about it, the check `((val & (APLIC_SRCCFG_SM)) != 2) && ((val & (APLIC_SRCCFG_SM)) != 3))` is also unnecessary, right? Because vaplic will never let it happen. I am not sure if I am missing something.